contracts/bulk-payments/src/bulk_payments.rs

```diff
@@ -0,0 +1,140 @@
+#![no_std]
+
+use multiversx_sc::derive_imports::*;
+use multiversx_sc::imports::*;
+
+const MAX_USERS_ALLOW: usize = 1000;
+const FOUR_HOURS: u64 = 60 * 60 * 4 * 1_000; // 4 hours
+
+#[type_abi]
+#[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, Clone, PartialEq)]
+pub struct UserAddrTimestamp<M: ManagedTypeApi> {
+    pub addr: ManagedAddress<M>,
+    pub timestamp: u64,
+}
+
+#[multiversx_sc::contract]
+pub trait BulkPayments {
+    #[init]
+    fn init(&self) {}
+
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[endpoint(optIn)]
+    fn opt_in(&self) {
+        let caller = self.blockchain().get_caller();
+        require!(
+            self.is_user_whitelisted(caller.clone()),
+            "User not whitelisted"
+        );
+
+        let user_already_opted_in = self.is_user_opted_in(caller.clone());
+
+        // If opted in and still getting rewards
+        if user_already_opted_in && !self.is_timestamp_expired(caller.clone()) {
```

Comment:
So you allow users who already opted in, but expired, to opt in again?

Reply:
Also, would it be helpful to add a time check for reentering? Or a maximum number of entries per epoch?

Reply:
For our use case we want to allow users who already opted in, but expired, to opt in again.
We check the time for reentering by calling `self.is_timestamp_expired(caller.clone()`.

For example: you `optIn` once and you'll receive rewards for the next 4h. In the mean time, `self.is_timestamp_expired(caller.clone()` will not let you `optIn` again.

After your timestamp expired, you can `optIn` again.